ocaml/xapi-storage-script/main.ml

```diff
@@ -1207,6 +1207,74 @@ let bind ~volume_script_dir =
     |> wrap
   in
   S.SR.scan sr_scan_impl ;
+  let sr_scan2_impl dbg sr =
+    Attached_SRs.find sr
+    >>>= (fun sr ->
+           return_volume_rpc (fun () -> Sr_client.stat (volume_rpc ~dbg) dbg sr)
+           >>>= fun response ->
+           Deferred.Result.return
+             {
+               Storage_interface.sr_uuid= response.Xapi_storage.Control.uuid
+             ; name_label= response.Xapi_storage.Control.name
+             ; name_description= response.Xapi_storage.Control.description
+             ; total_space= response.Xapi_storage.Control.total_space
+             ; free_space= response.Xapi_storage.Control.free_space
+             ; clustered= response.Xapi_storage.Control.clustered
+             ; health=
+                 ( match response.Xapi_storage.Control.health with
+                 | Xapi_storage.Control.Healthy _ ->
+                     Healthy
+                 | Xapi_storage.Control.Recovering _ ->
+                     Recovering
+                 | Xapi_storage.Control.Unreachable _ ->
+                     Unreachable
+                 | Xapi_storage.Control.Unavailable _ ->
+                     Unavailable
+                 )
+             }
+           >>>= fun sr_info ->
+           match sr_info.health with
+           | Healthy ->
+               return_volume_rpc (fun () ->
+                   Sr_client.ls
+                     (volume_rpc ~dbg ~compat_out:Compat.compat_out_volumes)
+                     dbg sr
+               )
+               >>>= fun response ->
+               let response = Array.to_list response in
+               (* Filter out volumes which are clone-on-boot transients *)
+               let transients =
+                 List.fold
+                   ~f:(fun set x ->
+                     match
+                       List.Assoc.find x.Xapi_storage.Control.keys
```

Comment:
I assume this list is short? Otherwise this could be costly if response is large.

Reply:
I copied this from the previous implementation, but `response` contains a list of volumes within an SR, and the max length of the list would be defined as the max num of volumes

Reply:
and this would depend on the size of `keys` as well, which I believe should be small. But even if it is large, there is not much we can do about it, creating an set for each `x` in the list of volumes? That would be a `O(mn)` operation itself.